File: api/src/main/java/net/blueshell/api/repository/EventRepository.java
package net.blueshell.api.repository;

import net.blueshell.api.base.BaseRepository;
import net.blueshell.api.model.File;
import net.blueshell.api.model.event.Event;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface EventRepository extends BaseRepository<Event> {

    @NotNull
    @Query("SELECT e FROM Event e ORDER BY e.startTime DESC")
    @Override
    Page<Event> findAll(@NotNull Pageable pageable);

    @NotNull
    @Query("SELECT e FROM Event e ORDER BY e.startTime DESC")
    @Override
    List<Event> findAll();
}

File: services/api/src/main/kotlin/net/blueshell/api/event/persistence/EventRepository.kt
package net.blueshell.api.event.persistence

import java.time.Instant

import org.springframework.data.repository.query.Param

import net.blueshell.api.shared.repository.BaseRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

@Repository
interface EventRepository : BaseRepository<Event, Long> {
    @EntityGraph(value = "Event.withBannerFileAndFormQuestions", type = EntityGraph.EntityGraphType.LOAD)
    override fun findAll(
        spec: Specification<Event>,
        pageable: Pageable,
    ): Page<Event>

    @EntityGraph(value = "Event.withBannerFileAndFormQuestions", type = EntityGraph.EntityGraphType.LOAD)
    override fun findAll(pageable: Pageable): Page<Event>

    @EntityGraph(value = "Event.withBannerFileAndFormQuestions", type = EntityGraph.EntityGraphType.LOAD)
    override fun findAll(): MutableList<Event>

    @Query(value = "SELECT * FROM events WHERE id = :id", nativeQuery = true)
    fun findByIdIncludingDeleted(id: Long): Event?

    @Query("select e.id from Event e where e.approved = true and e.startTime <= :to and e.endTime >= :from")
    fun findApprovedIdsOverlapping(
        @Param("from") from: Instant,
        @Param("to") to: Instant,
    ): List<Long>

    fun existsByTitle(title: String): Boolean

    @Query("select count(e) from Event e join e.gameCodes code where code = :code")
    fun countNamingGame(
        @Param("code") code: String,
    ): Long
}
